### ISSUE TITLE
Borrow compilation error with `vec.swap(0, vec.len() - 1)`, while `vec.push(vec.len())` works.

### ISSUE THREAD

**Issue by marcospb19** [rust-lang/rust#74319]
I tried compiling this code (here is the [playgound link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=041100f41b68065df13e143d640893a0)):
```rust
fn main() {
    let mut vec = vec![];
    
    vec.push(vec.len());
    vec.push(vec.len() - 1);    // This works!
    vec.swap(0, vec.len() - 1); // Error
}
```

I expected a  **Successful compilation**, but _this_ happened:

```
error[E0502]: cannot borrow `vec` as immutable because it is also borrowed as mutable
```
```rust
 --> src/main.rs:6:17
  |
6 |     vec.swap(0, vec.len() - 1);
  |     --- ----    ^^^ immutable borrow occurs here
  |     |   |
  |     |   mutable borrow later used by call
  |     mutable borrow occurs here
```

Note 1: that the line above the error also contains `vec.len()`, but it compiles.
Note 2: `vec.swap(0, vec.len());` produces the same error (nothing to do with ` - 1`).

### Meta

`rustc --version --verbose`:
```
rustc 1.44.1
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.44.1
LLVM version: 10.0
```

Also confirmed in the [playground](https://play.rust-lang.org/) against _Nightly_ and _Beta_.

**Comment by jonas-schievink** [comment]
two-phase borrows don't seem to work with autoderef at all, but I'm not sure if there's a canonical issue for that

**Comment by dlight** [comment]
Is autoderef happening in both of them or just in the vec.swap(..) case?

**Comment by jonas-schievink** [comment]
`push` is a method on `Vec`, so no autoderef is needed. `swap` is defined on slices, so one autoderef step is needed to call it.

**Comment by dlight** [comment]
`swap` is [defined on slices](https://doc.rust-lang.org/std/primitive.slice.html#method.swap) but [also on `Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.swap), so no autoderef should be needed.

**Comment by jonas-schievink** [comment]
No, that is in "Methods from `Deref<Target = [T]>`"

**Comment by nbdd0121** [comment]
That line is actually
```rust
<[_]>::swap(<Vec<_> as DerefMut>::deref_mut(&mut vec), 0, Vec::len(&vec) - 1);
```
And currently the `<Vec<_> as DerefMut>::deref_mut` would have to evaluate before `Vec::len`, which will fail borrowck.

**Comment by marcospb19** [comment]
I ran into the same issue again.

This did not compiled:
```rust
merge_sort(&mut v[..v.len()]);
merge_sort(&mut v[v.len()..]);
```

But this did:
```rust
let temp = v.len() / 2;
merge_sort(&mut v[..temp]);
merge_sort(&mut v[temp..]);
```

The signature:
```rust
fn merge_sort(v: &mut [i32]) { ... }
```
---

I'm out of context, are there any other other opened issues referencing this problem?

Will [Polonius](https://github.com/rust-lang/polonius) be able to fix it?

**Comment by nbdd0121** [comment]
> Will Polonius be able to fix it?

No. As I mentioned earlier, the desugared code will use a mutable borrow before an immutable one, so this code should not compile, otherwise it is unsound in general. To make NLL work across autoderef would likely be a langauge-level major change because the compiler'll need to move the location of autoderef.

**Comment by yume-chan** [comment]
This **doesn't** compile:

```rust
vec[vec.len() - 1] = 0;
```

But this **does**: ([playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=849f85cd45c5361647d3787269decf47))

```rust
use std::ops::IndexMut;
*vec.index_mut(vec.len() - 1) = 0;
```

The [doc of `IndexMut`](https://doc.rust-lang.org/std/ops/trait.IndexMut.html) says

> `container[index]` is actually syntactic sugar for `*container.index_mut(index)`, but only when used as a mutable value.

No, clearly not.

---

But this **doesn't** compile

```rust
*vec.get_mut(vec.len() - 1].unwrap() = 0;
```

**Nor** this:

```rust
*std::ops::IndexMut::index_mut(&mut vec, vec.len() - 1) = 0;
```

---

What's the difference?

**Comment by nbdd0121** [comment]
> No, clearly not, there is magic between them.

`container[index]` is indeed just a syntactic sugar, but it does not have two-phase borrow enabled, while two-phase borrow is enabled for a plain method call. This is actually intentional.

https://github.com/rust-lang/rust/blob/96859dbaf6229f131fbd427a32aaa95d4f9cb132/compiler/rustc_typeck/src/check/place_op.rs#L376-L380

Related: #49434

**Comment by pnkfelix** [comment]
After reading this and related issues, and re-reading:

https://github.com/rust-lang/rust/blob/96859dbaf6229f131fbd427a32aaa95d4f9cb132/compiler/rustc_middle/src/ty/adjustment.rs#L135-L151

Here is a summary of my understanding of the situation:

 * Some people are claiming that `v[IDX] = E` (where `IDX` is an expression that calls a `&self`-method on `v`) is rejected, while its desugared form is accepted by two-phase borrows
 * Other people are responding that `v[IDX] = E` doesn't desugar to `*v.index_mut(IDX) = E` (despite what its documentation says); it actually desugars to `*IndexMut::index_mut(&mut v, IDX) = E`, and the latter **is** rejected by the borrow-checker today, regardless of two-phase borrows.
 * The docs do claim that desugarings that turn into `Foo::some_method(&mut x, ...)` syntax are one class of desugarings supported by two-phase borrows, but clearly `IndexMut` is not placed in that class today.
 
So, what is currently blocking support for `v[IDX] = E` (where `IDX` calls `&self` method on `v`)...?
 * In theory we could extend two-phase borrows to support arbitrary`foo(&mut v, expr_using_v)`. That is what is tracked by #49434
 * Or we could treat `IndexMut` as a special case.

I've already expressed concerns about trying to deliver fully generalized two-phase borrows without a proof of soundness; you can see those comments on #49434.

(If you want to see an example of *why* we are trying to be conservative here, maybe take a look at #56254, which led us to ratchet back some of two-phase borrows initial expressiveness because we weren't confident that it was compatible with the stacked-borrows model from that time.)

A lot has happened since those discussions; e.g. models have evolved. But we still don't have any proof of soundness of generalized two-phase borrows, as far as I know.

(Having said all that, you all might be able to talk me into treating `IndexMut` as a special case... at the very least, we could experiment with such a change.)
